polynote-kernel/src/main/scala/polynote/kernel/lang/python/PythonInterpreter.scala

```diff
@@ -176,32 +197,22 @@ class PythonInterpreter(val kernelContext: KernelContext) extends LanguageInterp
         jep.set("__polynote_cell__", cellName)
         jep.set("__polynote_globals__", globals)
         expandGlobals()
-        // all of this parsing is just so if the last statement is an expression, we can print the value like the repl does
-        // TODO: should probably just use ipython to evaluate it instead
-        jep.eval("__polynote_parsed__ = ast.parse(__polynote_code__, __polynote_cell__, 'exec').body\n")
-        val numStats = jep.getValue("len(__polynote_parsed__)", classOf[java.lang.Long])
 
-        if (numStats > 0) {
-          jep.eval("__polynote_last__ = __polynote_parsed__[-1]")
-          val lastIsExpr = jep.getValue("isinstance(__polynote_last__, ast.Expr)", classOf[java.lang.Boolean])
+        jep.eval("__polynote_parsed__ = ast.parse(__polynote_code__, __polynote_cell__, 'exec')\n")
 
-          val resultName = "Out"
-          val maybeModifiedCode = if (lastIsExpr) {
-            val lastLine = jep.getValue("__polynote_last__.lineno", classOf[java.lang.Integer]) - 1
-            val (prevCode, lastCode) = code.linesWithSeparators.toSeq.splitAt(lastLine)
+        val numStats = jep.getValue("len(__polynote_parsed__.body)", classOf[java.lang.Long])
 
-            (prevCode ++ s"$resultName = (${lastCode.mkString})\n").mkString
-          } else code
+        if (numStats > 0) {
+          jep.eval("__polynote_parsed__ = LastExprAssigner().visit(__polynote_parsed__)")
+          jep.eval("__polynote_parsed__ = ast.fix_missing_locations(__polynote_parsed__)")
+          jep.eval("__polynote_code__ = compile(__polynote_parsed__, '<ast>', 'exec')")
```

Comment:
Neat – can `compile` give any further errors from parse? Would be great if we could surface parse and compile errors as `CompileErrors`!

Reply:
Hmm, I'm not sure we'd get any additional errors from `compile`, Since it [looks like](https://docs.python.org/3/library/ast.html#ast.parse)`ast.parse` is already a call to `compile`. 

However, we should be able to handle any `SyntaxErrors` we get from either parsing or compiling as `CompileErrors` to cover our bases. 

Tracking: https://github.com/polynote/polynote/issues/301